### DISCUSSION TITLE
Create a test coverage sort of "epic" to increase the test coverage on different packages

### DISCUSSION THREAD

**Discussion by pacyL2K19** [twentyhq/twenty#5648] [Engineering vision]
Currently the test coverage on multiple packages is low coverage, this could lead to some PRs breaking currently working features or introducing but without a clear record on the root that led to the bug.

For example, measuring the coverage on the `twenty-server` package, one can clearly see that many files are missing tests or having less test coverage (see the screenshot below)

![Screenshot 2024-05-29 at 13 16 46](https://github.com/twentyhq/twenty/assets/61581306/1ae7316e-3862-4a47-a0fe-5275a9054044)

> Note to the above screenshot: I have excluded some files in the jest config file, but we can still see that the coverage is still low

Here is the `collectCoverageFrom` array used for the above screenshot:
```
  collectCoverageFrom: [
    '**/*.(t|j)s',
    '!**/node_modules/**',
    '!**/*.input.(t|j)s',
    '!**/*.entity.(t|j)s',
    '!**/*.strategy.(t|j)s',
    '!**/*.args.(t|j)s',
    '!**/*.types.(t|j)s',
    '!**/*.entities.(t|j)s',
  ],
  ```

In other packages, one can see that many components (modules) miss test files

According to the [Linux Foundation's Recommended Practices for Hosting and Managing Open Source Projects on GitHub](https://www.linuxfoundation.org/research/hosting-os-projects-on-github) repport, `Continuous Testing and Integration` is one of the core open source principle that helps maintain a good collaborative work flow, is this something that we can work on in the future? As the project grows with time, this task can become time consuming to implement, so the sooner we implement it the better (I mean, if this is something is considered to be part of the workflow :))

This epic would consist on these steps:

- Measuring the current test coverage on each package
- Create sub-issues gradually and update the `coverageThreshold` as the coverage grows
- Add a test coverage report in the pipeline on PR creation so only PRs that don't destroy the current coverage pass, enforcing collaborators to implement tests to new features
- Following the [Testing Pyramid](https://semaphoreci.com/blog/testing-pyramid) principles, focus on writing unit tests, then add Integration and e2e tests on the most critical business logics that, if broken the system can be considered "half operational"

Kindly let me know if this is something that the maintainer's team considers implementing, I can help create sub-issues for each packages gradually with more details on each

@charlesBochet @FelixMalfait @thaisguigon @Bonapara

**Comment by charlesBochet** [comment]
@pacyL2K19 You are completely right. Adding tests to the server is on our top tier priority list. It's not easy to do because of our flexible schema. The current guideline is that all utils should be tested.
For services and controllers/resolvers, we would prefer having integration tests at this point but this needs to setup a database to do it and to reset it between tests. 
@magrinj will soon work on it!
